server/src/main/java/org/elasticsearch/action/bulk/MappingUpdatePerformer.java

```diff
@@ -23,11 +23,17 @@
 import org.elasticsearch.index.mapper.Mapping;
 import org.elasticsearch.index.shard.ShardId;
 
+import java.io.IOException;
+
 public interface MappingUpdatePerformer {
 
     /**
      * Update the mappings on the master.
      */
     void updateMappings(Mapping update, ShardId shardId, String type, ActionListener<Void> listener);
 
+    /**
+     * Validate the mapping update locally. Throws an unchecked exception of some kind if the mapping update is invalid.
+     */
+    void validateMappings(Mapping update, String type) throws IOException;
```

Comment:
Do we really need to increase the complexity of this interface to implement what we have now? 

It seems to me since we have a reference to `IndexShard primary` in `executeBulkItemRequest` we could simply inline the logic there and be happy since there's only one implementation anyway?

Reply:
Yeah that's how it was in the original commit, 1e3983c, but the `TransportShardBulkActionTests` have too much fakery in them for that. I don't like it either, I'll try harder to find another way to fix those tests.

Reply:
Ok, reverted in 006a8bf and achieved the same thing with mocks in 4b66099. This is ready for another look.